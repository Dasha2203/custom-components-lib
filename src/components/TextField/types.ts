import { InputHTMLAttributes } from 'react';

export type InputProps = InputHTMLAttributes<HTMLInputElement> & {
  label?: string;
  helperText?: string;
  error?: boolean;
};
