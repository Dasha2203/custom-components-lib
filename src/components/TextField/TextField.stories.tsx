import type { Meta, StoryObj } from '@storybook/react';

import TextField from './TextField';

const meta = {
  title: 'Input',
  component: TextField,
  tags: ['autodocs'],
} satisfies Meta<typeof TextField>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {},
};

export const Label = {
  args: {
    label: 'Label',
  },
};

export const Value = {
  args: {
    label: 'Label',
    value: 'Value',
  },
};

export const HelperText = {
  args: {
    label: 'Label',
    value: 'Value',
    helperText: 'Helper text',
  },
};

export const EmptyError = {
  args: {
    label: 'Label',
    error: true,
  },
};

export const ErrorWithValue = {
  args: {
    label: 'Label',
    error: true,
    value: 'Value',
  },
};

export const HelperTextWithError = {
  args: {
    label: 'Label',
    error: true,
    value: 'Value',
    helperText: 'Helper text',
  },
};

export const Disabled = {
  args: {
    label: 'Label',
    value: 'Value',
    disabled: true,
  },
};

export const NumberInput = {
  args: {
    label: 'Label',
    type: 'number',
    value: '5',
  },
};

export const DateInput = {
  args: {
    label: 'Label',
    type: 'date',
    value: '5',
  },
};
