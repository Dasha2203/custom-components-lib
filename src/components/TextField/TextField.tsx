import React, { useState } from 'react';
import clsx from 'clsx';
import * as classes from './style.module.scss';
import { InputProps } from './types';

const TextField = ({
  error,
  helperText,
  label,
  value,
  onChange,
  ...props
}: InputProps) => {
  const [isFocused, setIsFocused] = useState(false);

  const handleFocus = () => setIsFocused(true);
  const handleBlur = () => setIsFocused(!!value);

  const inputClasses = clsx(classes.textField__input, {
    [classes.textFieldRoot_active]: isFocused || value,
  });

  const rootClasses = clsx(classes.textFieldRoot, {
    [classes.textFieldRoot_error]: error,
    [classes.textFieldRoot_active]: isFocused || value,
  });

  return (
    <div className={rootClasses} data-testid="text-field">
      <div className={classes.textField}>
        <input
          {...props}
          value={value}
          className={inputClasses}
          onFocus={handleFocus}
          onBlur={handleBlur}
          onChange={onChange}
        />
        <label className={classes.textField__label}>{label}</label>
        <fieldset aria-hidden="true" className={classes.fieldset}>
          <legend className={classes.fieldset__legend}>
            {label && <span>{label}</span>}
          </legend>
        </fieldset>
      </div>
      {helperText && (
        <p className={classes.textFieldRoot__helperText}>{helperText}</p>
      )}
    </div>
  );
};

export default TextField;
