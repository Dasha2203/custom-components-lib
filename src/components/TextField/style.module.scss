.textFieldRoot {
  --error-color: #d32f2f;
  --primary-color: #1976d2;

  display: inline-flex;
  flex-direction: column;
  vertical-align: top;

  .textField {
    position: relative;
    display: inline-flex;
    align-items: center;
    font-weight: 400;
    font-size: 1rem;
    line-height: 1.4375em;
    letter-spacing: 0.00938em;
    color: rgba(0, 0, 0, 0.87);
    cursor: text;
    border-radius: 4px;

    &__label {
      position: absolute;
      left: 0;
      top: 0;
      z-index: 1;
      padding: 0;
      display: block;
      max-width: calc(100% - 24px);
      color: rgba(0, 0, 0, 0.6);
      font-weight: 400;
      font-size: 1rem;
      line-height: 1.4375em;
      letter-spacing: 0.00938em;
      transform-origin: top left;
      white-space: nowrap;
      overflow: hidden;
      text-overflow: ellipsis;
      transition:
        color 200ms cubic-bezier(0, 0, 0.2, 1),
        transform 200ms cubic-bezier(0, 0, 0.2, 1),
        max-width 200ms cubic-bezier(0, 0, 0.2, 1);
      pointer-events: none;
      transform: translate(14px, 16px) scale(1);
    }

    &__input {
      margin: 0;
      padding: 16.5px 14px;
      height: 1.4375em;
      width: 100%;
      font: inherit;
      letter-spacing: inherit;
      color: currentColor;
      border: 0;
      box-sizing: content-box;
      background: none;
      outline: none;

      &:disabled {
        color: rgba(0, 0, 0, 0.38);
        & ~ .fieldset {
          border-color: rgba(0, 0, 0, 0.38);
        }
        & ~ .textField__label {
          color: rgba(0, 0, 0, 0.38);
        }
      }

      &:focus {
        & ~ .fieldset {
          border-color: var(--primary-color);
          border-width: 2px;
        }
        & ~ .textField__label {
          color: var(--primary-color);
        }
      }
    }

    &:hover {
      &:has(.textField__input:not(:focus):not(:disabled)) {
        .fieldset {
          border-color: rgba(0, 0, 0, 0.87);
        }
      }
    }
  }

  &__helperText {
    margin: 3px 14px 0;
    font-weight: 400;
    font-size: 0.75rem;
    line-height: 1.66;
    letter-spacing: 0.03333em;
    text-align: left;
    color: rgba(0, 0, 0, 0.6);
  }
}

.textFieldRoot_active {
  .textField {
    &__input {
      & ~ .fieldset .fieldset__legend {
        max-width: 100%;
      }
      .fieldset__legend {
        max-width: 100%;
      }
      & ~ .textField__label {
        transform: translate(14px, -9px) scale(0.75);
        transform-origin: top left;
        pointer-events: auto;
      }
    }
  }
}

.textFieldRoot_error {
  .fieldset {
    border-color: var(--error-color);
  }

  .textField {
    &__input {
      &:focus {
        & ~ .fieldset {
          border-color: var(--error-color);
          border-width: 2px;
        }
        & ~ .textField__label {
          color: var(--error-color);
          transform: translate(14px, -9px) scale(0.75);
          transform-origin: top left;
          pointer-events: auto;
        }
      }
      & ~ .textField__label {
        color: var(--error-color);
      }
    }

    &:hover:has(.textField__input:not(:focus):not(:disabled)) {
      .fieldset {
        border-color: var(--error-color);
      }
    }
  }

  .textFieldRoot__helperText {
    color: var(--error-color);
  }
}

.fieldset {
  position: absolute;
  bottom: 0;
  right: 0;
  top: -5px;
  left: 0;
  margin: 0;
  padding: 0 8px;
  min-width: 0%;
  text-align: left;
  pointer-events: none;
  border-radius: inherit;
  border-style: solid;
  border-width: 1px;
  overflow: hidden;
  border-color: rgba(0, 0, 0, 0.23);

  &__legend {
    overflow: hidden;
    visibility: hidden;
    padding: 0;
    display: block;
    max-width: 0.01px;
    height: 11px;
    font-size: 0.75em;
    white-space: nowrap;
    transition: max-width 50ms cubic-bezier(0, 0, 0.2, 1);
    span {
      padding-inline: 5px;
      display: inline-block;
      opacity: 0;
      visibility: visible;
    }
  }
}
