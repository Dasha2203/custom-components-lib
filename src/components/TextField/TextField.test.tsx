import '@testing-library/jest-dom';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import TextField from './TextField';

describe('Testing TextField component', () => {
  test('Should render TextField component', () => {
    render(<TextField />);
    expect(screen.getByTestId('text-field')).toBeInTheDocument();
  });

  test('Should contain value to input', () => {
    const value = 'Some value';
    render(<TextField type="text" defaultValue={value} />);
    const input = screen.getByRole('textbox');

    expect(input).toBeInTheDocument();
    expect(input).toHaveValue(value);
  });

  test('Should set new value to input', async () => {
    const value = 'Some new value';
    render(<TextField type="text" />);
    const input = screen.getByRole('textbox');

    fireEvent.input(input, {
      target: { value },
    });

    expect(input).toBeInTheDocument();
    await waitFor(() => expect(input).toHaveValue(value));
  });

  test('Should contain "disabled" attribute', async () => {
    const initialValue = 'Initial value';
    const newValue = 'Some new value';

    render(<TextField type="text" disabled value={initialValue} />);
    const input = screen.getByRole('textbox');

    expect(input).toBeDisabled();
    expect(input).toHaveValue(initialValue);

    fireEvent.input(input, { target: { value: newValue } });

    await waitFor(() => expect(input).toHaveValue(initialValue));
  });

  test('Should show helper text', () => {
    const helperText = 'Helper text';
    render(<TextField helperText={helperText} />);
    const helperElement = screen.getByText(helperText);

    expect(helperElement).toBeInTheDocument();
  });
});

describe('TextField Snapshots', () => {
  test('Default', () => {
    const { asFragment } = render(<TextField />);

    expect(asFragment()).toMatchSnapshot();
  });

  test('TextField with value and label', () => {
    const { asFragment } = render(
      <TextField defaultValue={'Value'} label={'Label'} />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test('TextField with error and helperTExt', () => {
    const { asFragment } = render(
      <TextField
        error
        defaultValue={'Value'}
        label={'Label'}
        helperText={'Helper text'}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
