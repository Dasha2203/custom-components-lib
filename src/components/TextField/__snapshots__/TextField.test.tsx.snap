// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField Snapshots Default 1`] = `
<DocumentFragment>
  <div
    class="textFieldRoot"
    data-testid="text-field"
  >
    <div
      class="textField"
    >
      <input
        class="textField__input"
        value=""
      />
      <label
        class="textField__label"
      />
      <fieldset
        aria-hidden="true"
        class="fieldset"
      >
        <legend
          class="fieldset__legend"
        />
      </fieldset>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TextField Snapshots TextField with error and helperTExt 1`] = `
<DocumentFragment>
  <div
    class="textFieldRoot textFieldRoot_error"
    data-testid="text-field"
  >
    <div
      class="textField"
    >
      <input
        class="textField__input"
        value="Value"
      />
      <label
        class="textField__label"
      >
        Label
      </label>
      <fieldset
        aria-hidden="true"
        class="fieldset"
      >
        <legend
          class="fieldset__legend"
        >
          <span>
            Label
          </span>
        </legend>
      </fieldset>
    </div>
    <p
      class="textFieldRoot__helperText"
    >
      Helper text
    </p>
  </div>
</DocumentFragment>
`;

exports[`TextField Snapshots TextField with value and label 1`] = `
<DocumentFragment>
  <div
    class="textFieldRoot"
    data-testid="text-field"
  >
    <div
      class="textField"
    >
      <input
        class="textField__input"
        value="Value"
      />
      <label
        class="textField__label"
      >
        Label
      </label>
      <fieldset
        aria-hidden="true"
        class="fieldset"
      >
        <legend
          class="fieldset__legend"
        >
          <span>
            Label
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</DocumentFragment>
`;
